teps/0021-results-api.md

```diff
@@ -0,0 +1,663 @@
+integrated RBAC, a simple CLI (kubectl), generated client libraries, replicated
+on-cluster storage, and much more. But this also brings some limitations, which
+the project needs to work around to build a scalable and reliable CI/CD platform
+for users.
+
+### etcd
+
+All state described by a Kubernetes custom resource is stored, along with all
+Kubernetes cluster data, in etcd storage. etcd is a distributed key-value
+storage system that provides high availability and consistent reads and writes
+for Kubernetes API objects.
+
+Today, Tekton uses etcd to store the full execution history, which, if a user
+wants a permanent record of all previous executions, will eventually exhaust
+available storage. To avoid exhaustion, an operator can configure a garbage
+collection policy to delete old execution history to free up space on the
+cluster, but that means losing execution history. Storing execution history in
+the cluster also means that history disappears when the cluster disappears. Some
+cluster administrators prefer to think of clusters as cattle and not pets, and
+even accidental cluster deletions are a real possibility, so tying execution
+history to cluster storage can be problematic.
+
+etcd also provides no versioned history of Task and Pipeline objects, which
+might have been modified since they were run. This makes it impossible to know
+exactly what steps were executed by a previous run. There is work ongoing today
+to address this (versioned OCI image catalogs, storing the original TaskSpec in
+a TaskRunStatus, etc.)
+
+Tekton's API surface being limited to custom resource objects stored in a
+cluster's etcd also means that execution information not described as a CRD get
+dropped entirely. For example, triggering invocations (event received, k8s
+object created) are not stored in etcd, and so there's no way at all to query
+them.
+
+### Querying
+
+Being implemented as a set of Kubernetes custom resource objects stored in etcd,
+users can query the Kubernetes API server for the state of a single resource by
+name easily and efficiently. However, Kubernetes' support for listing/querying
+across many custom resources is severely lacking.
+
+Kubernetes allows users to request filtered lists of API resources using
+[field selectors and label selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/).
+For custom resources the only support field selectors are name and namespace.
+Label selectors are likewise limited, since we'd have to store any queryable
+information about a Run in a label for the Run to be able to query for it, and
+this grows large quickly. Together, these limitations make it pretty much
+impossible to query for anything actually useful today.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Define a Result API spec to store and query Tekton results.
```

Comment:
Added some definition here, as well as some discussion in Alternatives for how this relates to Task output results.

This is intentionally vague so that we can store broad data about TaskRuns and PipelineRuns. This will generally include the entire TaskRun/PipelineRun object (e.g. both spec and status), but we also see this evolving to include other information (input events, post-run receipts of GitHub status updates / Cloud Event publishing).

This is intended to be extensible to allow for future types which could include the artifacts you mentioned, though we are not looking to introduce new types as part of this proposal.